File: ecommerce-api-legacy/src/middlewares/errorHandler.js
'use strict';

/**
 * Tratamento de erro centralizado.
 *
 * No projeto original cada callback tratava (ou ignorava) o próprio erro, e
 * vários `if (err)` simplesmente não existiam — a requisição ficava pendurada.
 * Aqui todo erro assíncrono chega em um lugar só.
 */

const { AppError } = require('./errors');

/** Envolve um handler async para que rejeições cheguem ao error handler. */
function asyncHandler(fn) {
    return (req, res, next) => Promise.resolve(fn(req, res, next)).catch(next);
}

function notFoundHandler(req, res) {
    res.status(404).send('Recurso não encontrado');
}

function buildErrorHandler(logger) {
    // A assinatura de 4 argumentos é o que marca o middleware como error handler.
    return (err, req, res, _next) => {
        if (err instanceof AppError) {
            logger.info(`Erro de domínio (${err.status}): ${err.message}`);
            return res.status(err.status).send(err.message);
        }
        // stack trace vai para o log; o cliente recebe apenas a mensagem genérica
        logger.error(`Erro não tratado em ${req.method} ${req.originalUrl}`, {
            message: err.message,
            stack: err.stack,
        });
        return res.status(500).send('Erro Interno');
    };
}

module.exports = { asyncHandler, notFoundHandler, buildErrorHandler };

File: task-manager-api/src/middlewares/error_handler.py
"""Tratamento de erro centralizado.

O projeto original repetia `try/except` em cada rota — vários deles com `except:`
sem tipo, que engole inclusive KeyboardInterrupt e erros de programação.
"""
import logging

from flask import jsonify

from src.middlewares.errors import AppError

logger = logging.getLogger(__name__)


def _resposta(mensagem, status):
    return jsonify({"error": mensagem}), status


def registrar_error_handlers(app):
    @app.errorhandler(AppError)
    def _erro_de_dominio(exc):
        logger.info("Erro de domínio (%s): %s", exc.status, exc.mensagem)
        return _resposta(exc.mensagem, exc.status)

    @app.errorhandler(404)
    def _rota_desconhecida(_):
        return _resposta("Recurso não encontrado", 404)

    @app.errorhandler(405)
    def _metodo_nao_permitido(_):
        return _resposta("Método não permitido", 405)

    @app.errorhandler(Exception)
    def _erro_inesperado(exc):
        # stack trace no log; o cliente recebe apenas a mensagem genérica
        logger.exception("Erro não tratado: %s", exc)
        return _resposta("Erro interno", 500)
